optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java

```diff
@@ -0,0 +1,371 @@
+                .ifExists(Period.class, equal(Exam::getPeriod, Function.identity()))
+                .penalize("Period duration too short", HardSoftScore.ONE_HARD, Exam::getTopicStudentSize);
+    }
+
+    //    // More seating required during a period in a room than available in that room.
+//    rule "roomCapacityTooSmall"
+//    when
+//    $period : Period()
+//    $room : Room($capacity : capacity)
+//    accumulate(
+//            Exam(period == $period, room == $room, $studentSize : topicStudentSize);
+//    $totalStudentSize : sum($studentSize);
+//    $totalStudentSize > $capacity
+//        )
+//    then
+//        scoreHolder.addHardConstraintMatch(kcontext, $capacity - $totalStudentSize);
+//    end
+    private Constraint roomCapacityTooSmall(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Exam.class)
+                .join(Period.class, equal(Exam::getPeriod, Function.identity()))
+                .join(Room.class, equal((exam, period) -> exam.getRoom(), Function.identity()))
+                .groupBy((exam, period, room) -> room.getCapacity(),
+                         ConstraintCollectors.sum((exam, period, room) -> exam.getTopicStudentSize()))
+                .filter((capacity, totalStudentSize) -> totalStudentSize > capacity)
+                .penalize("Room capacity too small",
+                          HardSoftScore.ONE_HARD,
+                          (capacity, totalStudentSize) -> totalStudentSize - capacity);
+    }
+
+    //    rule "periodPenaltyExamCoincidence"
+//    when
+//    $periodPenalty : PeriodPenalty(
+//            periodPenaltyType == PeriodPenaltyType.EXAM_COINCIDENCE,
+//            $leftTopic : leftTopic,
+//            $rightTopic : rightTopic
+//    )
+//    $leftExam : Exam(topic == $leftTopic, $leftPeriod : period, period != null)
+//    $rightExam : Exam(topic == $rightTopic, period != $leftPeriod, period != null)
+//    // Left and right don't share a student because those are filtered out in ExaminationImporter
+//    then
+//        scoreHolder.addHardConstraintMatch(kcontext,
+//            - ($leftTopic.getStudentSize() + $rightTopic.getStudentSize()));
+//    end
+    private Constraint periodPenaltyExamCoincidence(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(PeriodPenalty.class)
+                .filter(periodPenalty -> periodPenalty.getPeriodPenaltyType() == PeriodPenaltyType.EXAM_COINCIDENCE)
+                .join(Exam.class,
+                      equal(PeriodPenalty::getLeftTopic, Exam::getTopic))
+                .filter((periodPenalty, leftExam) -> leftExam.getPeriod() != null)
+                .join(Exam.class,
+                      equal((periodPenalty, exam) -> periodPenalty.getRightTopic(), Exam::getTopic),
+                      equal((periodPenalty, exam) -> exam.getPeriod(), Exam::getPeriod))
+                .filter((periodPenalty, leftExam, rightExam) -> rightExam.getPeriod() != null)
```

Comment:
This is no longer necessary, right. `leftExam.getPeriod()` already isn't `null`, and `rightExam.getPeriod()` is joined where equal.